### PR TITLE
Unify dependency iteration in `ResolverState::get_dependencies`

### PR DESCRIPTION
**Pull request by konstin**
Upstack PR: #4430

Split out from #4430 according to https://github.com/astral-sh/uv/pull/4430#discussion_r1650192338.

### CODE REVIEW

File: crates/uv-resolver/src/resolver/mod.rs [hunk 1]
```diff
@@ -953,31 +953,11 @@ impl<InstalledPackages: InstalledPackagesProvider> ResolverState<InstalledPackag
...
-                );
-
-                let dependencies = match dependencies {
-                    Ok(dependencies) => dependencies,
-                    Err(err) => {
-                        return Ok(Dependencies::Unavailable(
-                            UnavailableVersion::ResolverError(uncapitalize(err.to_string())),
-                        ));
-                    }
-                };
```
Comment (konstin) [inline crates/uv-resolver/src/resolver/mod.rs:965]: @charliermarsh Do you know what this was about? It's only in this branch (in the non-root package branch we short-circuit on error) and the errors we're handling here look like they should be raised.

Comment (charliermarsh) [reply crates/uv-resolver/src/resolver/mod.rs:965]: Why should they be raised? You don't think we should backtrack in these cases?

Comment (konstin) [reply crates/uv-resolver/src/resolver/mod.rs:965]: As far as i can tell, the current implementation means that if there is a direct dependency (a root package dependency) where `PubGrubSpecifier::try_from` fails or which uses a banned URL, we mark those dependencies as unavailable, so the root package becomes unavailable(?). For regular dependencies, we already raise those errors directly using `from_requirements(...)?`.

This code looks like an artifact to me, but going through the git history unfortunately didn't clear this up either.

Comment (charliermarsh) [reply crates/uv-resolver/src/resolver/mod.rs:965]: It looks like there are _some_ errors that we'd want to recover from though, like `InvalidVersion` or perhaps `PubGrubSpecifierError`?

Comment (konstin) [reply crates/uv-resolver/src/resolver/mod.rs:965]: I don't think they are recoverable, as far as i can tell they lead to the root package being rejected in the next step, so you get the same error except it's wrapped in a message about the root package being unavailable.

Comment (konstin) [reply crates/uv-resolver/src/resolver/mod.rs:965]: We also do this only for the root package, which looks wrong.

Comment (charliermarsh) [reply crates/uv-resolver/src/resolver/mod.rs:965]: Where are they rejected? Note that this does not return an error; it returns `Ok(Dependencies::Unavailable)`.

Comment (konstin) [reply crates/uv-resolver/src/resolver/mod.rs:965]: This code is unreachable (both error cases are raised earlier), so i'm removing it

**Pull request merged**